### ISSUE TITLE
Logistic regression tool crashes on certain data

### ISSUE THREAD

**Issue by trotor** [gtkfi/ArcSDM#98]
![image](https://user-images.githubusercontent.com/10844537/40220999-c7deabfe-5a83-11e8-8468-8c78c69d7b98.png)

Preparing reproduce steps.

**Comment by trotor** [comment]
![image](https://user-images.githubusercontent.com/10844537/40225538-fc9150be-5a91-11e8-8362-c2e75b3c9549.png)
Can't reproduce with AG 10.4.1 - will try 10.6 later

**Comment by trotor** [comment]
"...You are using a shapefile as your raster mask. Needs to be single value integer raster. "

For the tools to work... or to make it crash? I need to understand and study this problem better, will study more next week on spare day.

**Comment by tachyon-work** [comment]
Both...

Why are you even using a shapefile as a raster mask? Regardless of whether it’s “easier” to get the tools to work that way, it is a fundamental requirement that the raster mask is a raster - it’s in the name. The entire modelling process is mathematically dependent on it! All the tools depend on cell counting in some capacity, which requires cell alignment and extent to be identical. The only way to ensure this is to use a raster mask. There is no exception to this. If the tool doesn’t work with a raster mask, this doesn’t mean you can just switch to a shapefile mask to get it to work. It means you need to figure out how to get it to work with a raster mask.

**Comment by trotor** [comment]
Okay, as I am not Geographer or GIS specialist and perhaps that is why I always thought it as a "mask" for operations, not raster mask as "mask that is raster and is used in operations". Looking at ESRI documentation it says:

> "Setting an analysis mask means that processing will only occur on locations that fall within the mask, and all locations outside of it will be assigned to NoData in the output."

And further it says if it is shapefile (feature dataset) then:

> If the analysis mask is a feature dataset, it will internally be converted to a raster on execution. For this reason, you should take care to ensure that the Cell Size and Snap Raster are set appropriately for your analysis

and none of the actual calculation code we converted does anything with the mask. It is all used in internal ESRI provided operations. But of course I will check this and try this with raster mask. And will report back. Thanks for the info!

**Comment by tachyon-work** [comment]
I think that’s where the misunderstanding is.

The tools don’t directly use the mask in the calculations, but all those temp rasters it creates _are based on that mask_. And those temp files _do_ get used in calculations. Your demo data already has your integer grids to use in the modelling, but they had to be created from scratch from the vector data at some point. During that phase of the modelling process (turning vector data into distance or density grids for example), the mask would have had to be set to a raster to get consistent outputs. If you change your mask part way through the project, it means the temp grids will be inaccurate. It messes up the maths. This is not a negotiable part of the modelling process :)

**Comment by trotor** [comment]
It seems like it is not the issue here (crashing - wrong calculations I don't know)
![image](https://user-images.githubusercontent.com/10844537/40291777-b601c634-5ccf-11e8-83c3-cf179e7d7a4e.png)

**Comment by trotor** [comment]
Still managed to run it after changes in my environment. Rechecking things after I update new version to my test machine and github.

![image](https://user-images.githubusercontent.com/10844537/40296328-ffcf0988-5ce4-11e8-9575-1d36b6ac78c9.png)


Problems in parametring are really tedious to work out and I will leave it for later.

**Comment by trotor** [comment]
Ah sorry, deleting previous entry. Still testing

**Comment by trotor** [comment]
Same problem as with #44 - Addjoin_management  crashes...

**Comment by trotor** [comment]
Minimum crash reproduction:
![image](https://user-images.githubusercontent.com/10844537/40300474-fde20bb8-5cf1-11e8-9976-ecb1ca0bb673.png)


`
arcpy.MakeRasterLayer_management(r"C:\tmp\issue\rc_em_real", "Outras_lyr");
arcpy.AddJoin_management("Outras_lyr", "VALUE", r"C:\tmp\issue\rc_em_real_C.dbf", "CLASS");
`

**Comment by trotor** [comment]
![image](https://user-images.githubusercontent.com/10844537/40300790-f0f6d68a-5cf2-11e8-8efe-27ef96ab2d4a.png)

**Comment by trotor** [comment]
10.6:
![image](https://user-images.githubusercontent.com/10844537/40300868-2b9c9aa4-5cf3-11e8-9971-336a38064435.png)


And here is the demo data to test this:
[Issue.zip](https://github.com/gtkfi/ArcSDM/files/2022026/Issue.zip)

**Comment by trotor** [comment]
It doesn't crash on empty environment, but on this map does: 
[MPM_DemoFiles_work.zip](https://github.com/gtkfi/ArcSDM/files/2022692/MPM_DemoFiles_work.zip)

There must be something wrong with the environment? Or the map?

**Comment by trotor** [comment]
[MPM_DemoFiles.zip](https://github.com/gtkfi/ArcSDM/files/2022720/MPM_DemoFiles.zip)

**Comment by tachyon-work** [comment]
Ok, I had a look at the mxd in the MPM_DemoFiles.zip file (can't open MPM_DemoFiles_work.zip because it's saved in 10.6 format and is not backwards compatible with any version we have). There is no snap raster or mask set. I understand that you may have modified these to avoid the crash, but they need to be reset.

Regarding your comment about the ESRI documentation on the environment settings, that misses a fundamental issue - it regenerates the raster from the shapefile mask every time you execute a tool, and the point is that _it does not produce an identical raster alignment/coverage every time_. This is crucial for the modelling workflow. Yes, _technically_ the ESRI documentation is correct, but it is irrelevant in the context of the modelling workflow. The only way to ensure that the raster alignment and extent is identical for each processing job in the modelling workflow is to use a snap raster and raster mask the entire time.

**Comment by trotor** [comment]
I understand your point, thank you for clearing that. I will direct your comment to project members and document  it to wiki also.

However, single value integer mask works and the crash seems not to be dependant on it in this case (atleast running on 10.4.1).  There wasn't really simple way to calculate nodata containing raster size in ArcGis so I implemented new way to calculate it with numpy.  Now all kinds of raster masks should be ok.

On 10.5.1 and 10.6 using those mxds Arcgis does crash using only those two lines of code in python console window. ArcMap doesn't crash as spectacularly as when using tool, but background geoprocessing task crashes nevertheless. Unfortunately ESRI seems to be little reluctant to accept it as a bug. 

Issue #44 is basically the same and it is kind of still open, I never heard from ESRI. It was caused by coordinate transformations but this case seems to not have exactly that, but most likely something similar.

There is mxd version on later porst for 10.4. and after testing. I will edit the message to be more clear.
